faiss_vector_posting.go

```diff
@@ -395,156 +392,86 @@ func (sb *SegmentBase) InterpretVectorIndex(field string, requiresFiltering bool
 					if err != nil {
 						return nil, err
 					}
-
 					addIDsToPostingsList(rv, ids, scores)
 					return rv, nil
 				}
-
 				// vector IDs corresponding to the local doc numbers to be
 				// considered for the search
 				vectorIDsToInclude := make([]int64, 0, len(eligibleDocIDs))
 				for _, id := range eligibleDocIDs {
 					vectorIDsToInclude = append(vectorIDsToInclude, docVecIDMap[uint32(id)]...)
 				}
-
+				// In case a doc has invalid vector fields but valid non-vector fields,
+				// filter hit IDs may be ineligible for the kNN since the document does
+				// not have any/valid vectors.
 				if len(vectorIDsToInclude) == 0 {
 					return rv, nil
 				}
-
-				// Retrieve the mapping of centroid IDs to vectors within
-				// the cluster.
-				clusterAssignment, _ := vecIndex.ObtainClusterToVecIDsFromIVFIndex()
-				// Accounting for a flat index
-				if len(clusterAssignment) == 0 {
+				// If the index is not an IVF index, then the search can be
+				// performed directly, using the Flat index.
+				if !vecIndex.IsIVFIndex() {
+					// vector IDs corresponding to the local doc numbers to be
+					// considered for the search
 					scores, ids, err := vecIndex.SearchWithIDs(qVector, k,
 						vectorIDsToInclude, params)
 					if err != nil {
 						return nil, err
 					}
-
 					addIDsToPostingsList(rv, ids, scores)
 					return rv, nil
 				}
-
-				// Converting to roaring bitmap for ease of intersect ops with
-				// the set of eligible doc IDs.
-				centroidVecIDMap := make(map[int64]*roaring.Bitmap)
-				for centroidID, vecIDs := range clusterAssignment {
-					if _, exists := centroidVecIDMap[centroidID]; !exists {
-						centroidVecIDMap[centroidID] = roaring.NewBitmap()
-					}
-					vecIDsUint32 := make([]uint32, 0, len(vecIDs))
-					for _, vecID := range vecIDs {
-						vecIDsUint32 = append(vecIDsUint32, uint32(vecID))
-					}
-					centroidVecIDMap[centroidID].AddMany(vecIDsUint32)
-				}
-
 				// Determining which clusters, identified by centroid ID,
 				// have at least one eligible vector and hence, ought to be
 				// probed.
-				eligibleCentroidIDs := make([]int64, 0)
-
+				clusterVectorCounts, err := vecIndex.CountVectorsPerCluster(vectorIDsToInclude)
```

Comment:
Rename this, tandem comment in https://github.com/blevesearch/go-faiss/pull/41#pullrequestreview-2714000662